graphql-ballerina/src/graphql/parser/parser.bal

```diff
@@ -0,0 +1,208 @@
+// Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+//
+// WSO2 Inc. licenses this file to you under the Apache License,
+// Version 2.0 (the "License"); you may not use this file except
+// in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+class Parser {
+    private Lexer lexer;
+    private DocumentNode document;
+
+    public isolated function init(string text) returns ParsingError? {
+        self.lexer = new(text);
+        self.document = new;
+    }
+
+    public isolated function parse() returns DocumentNode|ParsingError {
+        check self.blockAnalysis();
+        check self.populateDocument();
```

Comment:
Can we go for a recursive descent parser?

That is more clean and you don't have to traverse the input multiple times.

Reply:
Updated